src/content/resources/javascript/promises/async-await.md

```diff
@@ -0,0 +1,137 @@
+---
+authors:
+  - "ddivad#1337"
+created_at: "2019/10/01"
+title: Async Await
+recommended_reading:
+  - javascript/promises
+---
+
+## Async Await
+
+Async / Await is another way to handle the calling of asynchronous code and is built on top of Promises. It is a nice and clean way to handle asynchronous code, without the need to `.then()` functions to the promise.
+
+It is made up of 2 keywords as the name suggests: `async` and `await`. These need to be used together and can for this method to work.
+
+## Async
+
+The `async` keyword is used to show that the function is going to return a promise. Any return values from the function will be converted to a promise automatically, if they are not already.
+
+```js
+async function foo() {
+  return "foo";
+}
+```
+
+This basic example will return a promise with the value of `foo`.
+
+## Await
+
+The `await` keyword is used to wait until a promise has executed and fetches the result. In order to use the `await` keyword, you **need** to be inside an `async` function. 
+
+```js
+let data = await foo();
+console.log(foo); // "foo"
+```
+
+This code will call the `foo()` function from above and will wait on that line until the promise returned from the `async` function has returned.
+
+Another example to demonstrate this is if we extended the `foo()` function from above to add a 5 second delay.
+
+```js
+async function foo() {
+  let promise = new Promise((resolve, reject) => {
+    setTimeout(() => resolve("foo"), 5000)
+  });
+
+  let data = await promise; // Execution waits here until promise resolves
+  console.log(data); // "foo"
+}
+```
+
+## Error Handling
+
+As with Promises, if the Promise returns with an error, it will affect the `await` call that triggered it. When using Promises without Async / Await, the `.catch()` syntax is used, and this also exists with Aysnc / Await using the `try...catch` syntax. This can be demonstrated by making our `foo()` function throw an error.
+
+```js
+async function foo() {
+  throw new Error("An Error Occurred");
+}
+```
+
+This would be handled using Aysnc / Await like so:
+
+```js
+async function getData() {
+  try {
+    let data = await foo();
+    console.log(data);
+  } catch (err) {
+    console.error(err); // err: An Error Occurred
+  }
+}
+```
+
+## Beginner Mistakes
+
+As mentioned in the [intro](./intro.md), Promises are the highest source of confusion for beginners. Async / Await adds another layer on top of Promises, and comes with its own pitfalls.
+
+`await` will **only** work if the function you try to add `await` to is `async`. Using `await` in top level code will not work (yet).
+
+```js
+let data = await foo(); // syntax error
+```
+
+To get around this issue, you can use an declare asynchronous function anonymously, if needed:
+
+```js
+(async () => {
+  let data = await foo();
+  console.log(data); // "foo"
+})();
```

Comment:
Might be a good idea to to add a single link to the IIFE you're using here.

Reply:
I'm not sure what you mean here? Do you mean add a link to what an IIFE is?

Reply:
Yeah, either that or just turn it into a regular function call. I think we probably want to keep the examples as simple as possible in every way

Reply:
Yea, I get what you're saying, but to me an IIFE here makes sense as people can copy / paste the code and it will just work. If it was a normal function instead, they would have to make an async function themselves if they wanted to run the code and check the output, which was the reason I used it here. I'll see if I can find a good link to put in here maybe that can explain them, or maybe even write an explanation in some other article here for how they work.